codice/ddf
# DDF-3573 Create FeatureTransformer API and Implementation

What does this PR do?
This PR refactors the existing WFS code to remove some of the XStream-specific logic used to create metacards. It also introduces a templating capability to map and transform WFS response element values to metacard values.

Who is reviewing it?
@troymohl
@mojogitoverhere
@BernardIgiri
@GabrielFabian

(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)

Select relevant component teams:
https://github.com/orgs/codice/teams

Choose 2 committers to review/merge the PR.
(please choose ONLY two committers from below, delete the rest)
@jlcsmith
@pklinef

How should this be tested? (List steps with links to updated documentation)
Any background context you want to provide?

Testing with default mappings:

1) In the admin console, create a "WFS v1.1.0 Federated Source"
2) In Intrigue, execute a search
   - verify that the metacards contain the attributes and values returned by the WFS server

Testing with templated mappings:

1) In the admin console, create a "WFS v1.1.0 Federated Source"
2) In the admin console, create a "WFS Feature to Metacard Templated Mapping"
   a) enter the Feature Type, e.g. {http://www.namespace.example}LocalName
   b) enter one or more attribute mappings in JSON format, e.g. {attributeName:"ext.title", featureName:"state_name", template:"State: {{state_name}}"}
3) In Intrique, execute a search
   - verify that the metacards contain the attributes and values mapped in step #2

What are the relevant tickets?

DDF-3573

Screenshots (if appropriate)
Checklist:
[-] Documentation Updated
[X] Update / Add Unit Tests
[-] Update / Add Integration Tests
Notes on Review Process
Please see Notes on Review Process for further guidance on requirements for merging and abbreviated reviews.

Review Comment Legend:
:pencil2:  (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist.
:question:  (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
:exclamation:  (Exclamation Mark) This comment is critical and requires clarification or action before approval.
